pkg/backoff/backoff.go

```diff
@@ -0,0 +1,123 @@
+
+	"github.com/spf13/viper"
+)
+
+const (
+	// resetAfter sets the total seconds a backoff will last.
+	resetAfter = 3600
+	// factor is the total seconds to be multiplied by.
+	factor = 15
+)
+
+// RetryConfig defines backoff data.
+type RetryConfig struct {
+	// At is the time when the first failure happened.
+	At *time.Time
+	// Retries is the number of attempts to connect.
+	Retries int
+	// V is an instance of Viper.
+	V *viper.Viper
+}
+
+// WithRetry initializes and returns a heartbeat handle option, which
+// can be used in a heartbeat processing pipeline to prevent trying to send
+// a heartbeat when the api is unresponsive.
+func WithRetry(config RetryConfig) heartbeat.HandleOption {
+	return func(next heartbeat.Handle) heartbeat.Handle {
+		return func(hh []heartbeat.Heartbeat) ([]heartbeat.Result, error) {
+			log.Debugln("execute heartbeat backoff algorithm")
+
+			nextDuration, _ := time.ParseDuration(fmt.Sprintf("%fs", float64(factor)*math.Pow(2, float64(config.Retries))))
+
+			now := time.Now()
+			if config.At != nil && now.Before(config.At.Add(nextDuration)) {
+				log.Debugf("current time '%s' is before next backoff at '%s'",
+					now.Format(time.Stamp),
+					config.At.Add(nextDuration).Format(time.Stamp))
+
+				return nil, fmt.Errorf("won't send heartbeat due to backoff")
+			}
+
+			results, err := next(hh)
+			if err != nil {
+				log.Debugf("incrementing backoff due to error")
+
+				retries := config.Retries + 1
+
+				// on error update backoff settings
+				if updateErr := updateBackoffSettings(config.V, retries, config.At); updateErr != nil {
+					log.Warnf("failed to update backoff settings: %s", updateErr)
+				}
+
+				return nil, err
+			}
+
+			if config.At != nil {
+				// no error returned then reset backoff settings
+				if resetErr := resetBackoffSettings(config.V); resetErr != nil {
+					log.Warnf("failed to reset backoff settings: %s", resetErr)
+				}
+			}
+
+			return results, nil
+		}
+	}
+}
+
+func updateBackoffSettings(v *viper.Viper, retries int, at *time.Time) error {
+	w, err := ini.NewIniWriter(v, ini.FilePath)
+	if err != nil {
+		return fmt.Errorf("failed to parse config file: %s", err)
+	}
+
+	keyValue := map[string]string{
+		"backoff_retries": strconv.Itoa(retries),
+	}
+
+	now := time.Now()
+
+	if at == nil {
+		keyValue["backoff_at"] = now.Format(ini.DateFormat)
+	}
+
+	// reset backoff settings if reached max allowed time
+	if at != nil && now.Sub(*at).Seconds() > resetAfter {
+		keyValue["backoff_at"] = now.Format(ini.DateFormat)
+		keyValue["backoff_retries"] = "1"
+	}
+
+	if err := w.Write("internal", keyValue); err != nil {
+		return fmt.Errorf("failed to write to config file: %s", err)
+	}
+
+	return nil
+}
+
+func resetBackoffSettings(v *viper.Viper) error {
+	w, err := ini.NewIniWriter(v, ini.FilePath)
+	if err != nil {
+		return fmt.Errorf("failed to parse config file: %s", err)
+	}
+
+	keyValue := map[string]string{
+		"backoff_at":      time.Now().Format(ini.DateFormat),
```

Comment:
We reset here. Reset ends the backoff, thus after resetting in `WithRetry`, `config.At` should be `nil` to not cause any backoff.

But here we set `backoff_at` to a timestamp. `config.At` won't be `nil` in `WithRetry` then, or am I missing something?

Reply:
To get backoff reset means no error has returned from the API, thus the process will exit with zero which means no retry will be done at current process. Tne next execution will load again the params and go on..

Reply:
Latest changes fixed the problem.